queasars/job_shop_scheduling/domain_wall_hamiltonian_encoder.py

```diff
@@ -0,0 +1,528 @@
+                domain_wall_variable = self._operation_start_variables[operation]
+                start_time = domain_wall_variable.value_from_bitlist(bit_list=bit_list)
+                if start_time is not None:
+                    scheduled_operations.append(ScheduledOperation(operation=operation, start_time=start_time))
+                else:
+                    scheduled_operations.append(UnscheduledOperation(operation=operation))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self.jssp_instance, schedule=job_schedules)
+
+    def _prepare_encoding(self) -> None:
+        """Counts the needed qubits to encode the problem" and assigns the necessary domain wall variables"""
+        for job in self.jssp_instance.jobs:
+            for i, operation in enumerate(job.operations):
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                start_offset = sum(operation.processing_duration for j, operation in enumerate(job.operations) if j < i)
+                end_offset = sum(operation.processing_duration for j, operation in enumerate(job.operations) if j >= i)
+
+                n_start_times = self.time_limit - (start_offset + end_offset) + 1
+
+                if n_start_times < 1:
+                    raise ValueError("There is no feasible solution for the given time_limit!")
+
+                self._operation_start_variables[operation] = DomainWallVariable(
+                    qubit_start_index=self._n_qubits,
+                    values=tuple(range(start_offset, start_offset + n_start_times)),
+                )
+
+                self._n_qubits += self._operation_start_variables[operation].n_qubits
+
+        self._encoding_prepared = True
+
+    def _prepare_hamiltonian(self):
+        """
+        Gathers the terms making up the hamiltonian. These include penalties for invalid variable states,
+        penalties for invalid ordering and overlapping of operations and an optimization term to minimize the
+        makespan of the problem
+        """
+        for job in self.jssp_instance.jobs:
+            for operation in job.operations:
+                variable_viability_term = self._operation_start_variables[operation].viability_term(
+                    penalty=self._encoding_penalty, quantum_circuit_n_qubits=self._n_qubits
+                )
+                self._local_terms.append(variable_viability_term)
+
+            for i in range(0, len(job.operations) - 1):
+                precedence_term = self._operation_precedence_term(
+                    job.operations[i], job.operations[i + 1], self._constraint_penalty
+                )
+                self._local_terms.extend(precedence_term)
+
+        for operations in self._machine_operations.values():
```

Comment:
As in the SCIP solver, this is a list of operations corresponding to a machine. I would prefer something like `for _, operations in self._machine_operations.items()` or rename the `operations` variable to make this a bit easier to understannd.